fix #49754 minion zmq connecting to master configured explicitly as IPv6 address or IPv6 only master

aphor opened the pull request:
### What does this PR do?
fixes implementation of IPv6 address handling by minion and zeromq transport to enable minions to connect to a master identified only by IPv6 address or address/port

Work:
  - depends on #49705 [MERGED]
  - unit tests failing [FIXED]
  - integration tests failing [FIXED]
  - parsing config master: "host:port" [FIXED]
  - missing coverage for host:port parsing [FIXED]

### What issues does this PR fix or reference?
#49754, possibly #40515

### Tests written?

Yes, or rather test coverage added to existing tests.

### Commits signed with GPG?

Yes

aphor committed:
fix minion zmq connecting to master configured as IPv6 address

aphor committed:
fix minion zmq connecting to master configured as IPv6 address

isbm commented on salt/minion.py:
```diff
@@ -240,27 +240,39 @@ def resolve_dns(opts, fallback=True):
     # definition such as master: '[::1]:1234'.
     if opts['master_uri_format'] == 'ip_only' or salt.utils.network.is_ipv6(opts['master']):
-        ret['master'] = opts['master']
+        ret['master'] = str(ipaddress.ip_address(opts['master']))
```
Just to note here that https://github.com/saltstack/salt/pull/49705 will add scope on IPv6, if any, once you `__str__` the object. That said, if  you have something like `fe80::1ff:fe23:4567:890a%eth2` in  as an input, you will get _exactly_ the same thing back.

And vice versa, without mentioned PR, if you have scope, now it will just crash (all python versions affected from 2.6 to 3.7).

aphor replied:
I submitted an upstream bug for ipaddress.IPv6Address() scoped link local addresses for a long term fix (or won't fix, if that's all we can get).

I think I want to make this PR depend on #49705 though.

isbm replied:
@aphor yeah, your should go after. I would expect some breakage, you should be able to resolve it though, as it is just imports in your case.

isbm commented on salt/transport/zeromq.py:
```diff
@@ -69,33 +69,38 @@ def _get_master_uri(master_ip,
+                master_ip_address = ipaddress.ip_address(master_ip)
+                if master_ip_address.Version == 4:
+                    ip_any = '0.0.0.0'
+                if master_ip_address.Version == 6:
```
You should use `.version` property instead (lower-case). And since there is only two versions of IP, you can just use `else` instead. :wink:

isbm commented on salt/utils/win_reg.py:
```diff
@@ -697,17 +701,20 @@ def set_value(hive,
         log.exception('Encountered error setting registry value')
         return False
+    finally:
+        if handle:
```
Please check for `None` explicitly:

```python
if handle is not None:
   ...
```

isbm commented on salt/utils/win_reg.py:
```diff
@@ -938,3 +945,6 @@ def delete_value(hive, key, vname=None, use_32bit_registry=False):
             log.error('32bit Reg: %s', use_32bit_registry)
             return False
+    finally:
+        if handle:
```
Please explicitly check for `None` here (see above comment).

isbm commented on tests/integration/utils/testprogram.py:
```diff
@@ -416,7 +419,10 @@ def detach_from_parent_group():
             popen_kwargs['preexec_fn'] = detach_from_parent_group
 
-        self.argv = [self.program]
+        if salt.utils.platform.is_windows():
```
If you hit a very strange Windows machine, this might crash "program not found". Therefore I would also check `$PATH` if `python.exe` is also around.

aphor replied:
I messed up rebasing my commits on 2018.3 and accidentally pulled in a bunch of other commits. I'm going to try and fix that and force push the minimized changes back to this PR branch.

KaiSforza commented on salt/utils/zeromq.py:
```diff
@@ -82,6 +82,8 @@ def ip_bracket(addr):
     '''
-    if addr and ':' in addr and not addr.startswith('['):
-        return '[{0}]'.format(addr)
+    _ip_addr = ipaddress.ip_address(addr)
```
The `ipaddress` module isn't imported for use here.

isbm replied:
Oh, right. And it also should be imported from `_compat`, not directly!

KaiSforza replied:
It's also missing from the `salt/transport/zeromq.py` file.

KaiSforza commented on salt/utils/zeromq.py:
```diff
@@ -82,6 +82,8 @@ def ip_bracket(addr):
+    _ip_addr = ipaddress.ip_address(addr)
+    if _ip_addr.version == 6:
+        return '[{0}]'.format(str(_ip_addr))
+    log.trace("found zeromq transport addr:{0} using addr:{}".format(raw, addr))
```
`raw` isn't defined, also you should be using `{0}` and `{1}`, not an empty bracket.

aphor replied:
I'm fixing this stuff. Coffee helps :)

rallytime commented:
Hrm. The tests aren't liking this change very much. There are quite a few failing tests here.

aphor commented:
This adds coverage for IPv6 connectivity. I'll check things out after a
rebate.

On Thu, Oct 4, 2018, 08:01 Nicole Thomas <notifications@github.com> wrote:

> Hrm. The tests aren't liking this change very much. There are quite a few
> failing tests here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/saltstack/salt/pull/49755#issuecomment-427009529>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAOKuw9AfvVrqGobmeZsXNPnT7AaRqR4ks5uhgbEgaJpZM4W2FcG>
> .
>

aphor commented:
> Hrm. The tests aren't liking this change very much. There are quite a few failing tests here.

I rebased against 2018.3 head last night and only got 2 unit test failures.

This one(`tests/unit/transport/test_zeromq.py", line 335`) needs some work, so I'll look at that.

```
 --------  Failed Tests  ----------------------------------------------------------------------------------------------------------------------------------------------------------------
   -> unit.test_doc.DocTestCase.test_check_for_doc_inline_markup  .......................................................................................................................
       Traceback (most recent call last):
         File "/Users/aphor/src/salt.git/tests/unit/test_doc.py", line 85, in test_check_for_doc_inline_markup
           self.assertEqual(test_ret, {})
       AssertionError: {'/Users/aphor/src/salt.git//lib/python3.6[2365 chars]n.']} != {}
       + {}
       - {'/Users/aphor/src/salt.git//lib/python3.6/site-packages/astroid/nodes.py': [':doc:`ast '
       -                                                                              'documentation '
       -                                                                              '<green_tree_snakes:nodes>`'],
       -  '/Users/aphor/src/salt.git//lib/python3.6/site-packages/botocore/docs/bcdoc/style.py': ["self.doc.write(':doc:`%s "
       -                                                                                          "<%s>`' "
       -                                                                                          '% '
       -                                                                                          '(title, '
       -                                                                                          'link))',
       -                                                                                          "self.doc.write(':doc:`%s "
       -                                                                                          "<%s>`' "
       -                                                                                          '% '
       -                                                                                          '(title, '
       -                                                                                          'page))'],
       -  '/Users/aphor/src/salt.git//lib/python3.6/site-packages/docker/client.py': [':doc:`configs '
       -                                                                              'documentation '
       -                                                                              '<configs>` '
       -                                                                              'for '
       -                                                                              'full '
       -                                                                              'details.',
       -                                                                              ':doc:`containers '
       -                                                                              'documentation '
       -                                                                              '<containers>` '
       -                                                                              'for '
       -                                                                              'full '
       -                                                                              'details.',
       -                                                                              ':doc:`images '
       -                                                                              'documentation '
       -                                                                              '<images>` '
       -                                                                              'for '
       -                                                                              'full '
       -                                                                              'details.',
       -                                                                              ':doc:`networks '
       -                                                                              'documentation '
       -                                                                              '<networks>` '
       -                                                                              'for '
       -                                                                              'full '
       -                                                                              'details.',
       -                                                                              ':doc:`nodes '
       -                                                                              'documentation '
       -                                                                              '<nodes>` '
       -                                                                              'for '
       -                                                                              'full '
       -                                                                              'details.',
       -                                                                              ':doc:`plugins '
       -                                                                              'documentation '
       -                                                                              '<plugins>` '
       -                                                                              'for '
       -                                                                              'full '
       -                                                                              'details.',
       -                                                                              ':doc:`secrets '
       -                                                                              'documentation '
       -                                                                              '<secrets>` '
       -                                                                              'for '
       -                                                                              'full '
       -                                                                              'details.',
       -                                                                              ':doc:`services '
       -                                                                              'documentation '
       -                                                                              '<services>` '
       -                                                                              'for '
       -                                                                              'full '
       -                                                                              'details.',
       -                                                                              ':doc:`swarm '
       -                                                                              'documentation '
       -                                                                              '<swarm>` '
       -                                                                              'for '
       -                                                                              'full '
       -                                                                              'details.',
       -                                                                              ':doc:`volumes '
       -                                                                              'documentation '
       -                                                                              '<volumes>` '
       -                                                                              'for '
       -                                                                              'full '
       -                                                                              'details.'],
       -  '/Users/aphor/src/salt.git//lib/python3.6/site-packages/nose/plugins/__init__.py': [':doc:`IPluginInterface '
       -                                                                                      '<interface>` '
       -                                                                                      'in '
       -                                                                                      'nose.plugins.base. '
       -                                                                                      'Please '
       -                                                                                      'note '
       -                                                                                      'that',
       -                                                                                      'into '
       -                                                                                      'the '
       -                                                                                      ':doc:`IPluginInterface '
       -                                                                                      '<interface>` '
       -                                                                                      'to '
       -                                                                                      'see '
       -                                                                                      'all '
       -                                                                                      'available '
       -                                                                                      'hooks.',
       -                                                                                      'Subclass '
       -                                                                                      ':doc:`ErrorClassPlugin '
       -                                                                                      '<errorclasses>`.',
       -                                                                                      'Examples: '
       -                                                                                      ':doc:`nose.plugins.deprecated '
       -                                                                                      '<deprecated>`,',
       -                                                                                      ':doc:`nose.plugins.skip '
       -                                                                                      '<skip>`',
       -                                                                                      'Examples: '
       -                                                                                      ':doc:`nose.plugins.capture '
       -                                                                                      '<capture>`,',
       -                                                                                      ':doc:`nose.plugins.failuredetail '
       -                                                                                      '<failuredetail>`',
       -                                                                                      'Example: '
       -                                                                                      ':doc:`nose.plugins.doctests '
       -                                                                                      '<doctests>`',
       -                                                                                      'Examples: '
       -                                                                                      ':doc:`nose.plugins.cover '
       -                                                                                      '<cover>`, '
       -                                                                                      ':doc:`nose.plugins.prof '
       -                                                                                      '<prof>`',
       -                                                                                      'Examples: '
       -                                                                                      ':doc:`nose.plugins.attrib '
       -                                                                                      '<attrib>`,',
       -                                                                                      ':doc:`nose.plugins.doctests '
       -                                                                                      '<doctests>`, '
       -                                                                                      ':doc:`nose.plugins.testid '
       -                                                                                      '<testid>`'],
       -  '/Users/aphor/src/salt.git//lib/python3.6/site-packages/nose/plugins/allmodules.py': ['More '
       -                                                                                        'information: '
       -                                                                                        ':doc:`../doc_tests/test_allmodules/test_allmodules`',
       -                                                                                        'the '
       -                                                                                        ':doc:`doctest '
       -                                                                                        'plugin '
       -                                                                                        '<doctests>`. '
       -                                                                                        'This '
       -                                                                                        'is '
       -                                                                                        'because '
       -                                                                                        'any '
       -                                                                                        'given'],
       -  '/Users/aphor/src/salt.git//lib/python3.6/site-packages/nose/plugins/doctests.py': ['See '
       -                                                                                      ':doc:`../doc_tests/test_doctest_fixtures/doctest_fixtures` '
       -                                                                                      'for'],
       -  '/Users/aphor/src/salt.git//lib/python3.6/site-packages/nose/plugins/multiprocess.py': ['See '
       -                                                                                          ':doc:`../doc_tests/test_multiprocess/multiprocess` '
       -                                                                                          'for'],
       -  '/Users/aphor/src/salt.git//lib/python3.6/site-packages/tornado/web.py': ['See '
       -                                                                            'the '
       -                                                                            ':doc:`guide` '
       -                                                                            'for '
       -                                                                            'additional '
       -                                                                            'information.']}
   ......................................................................................................................................................................................
   -> unit.transport.test_zeromq.ZMQConfigTest.test_master_uri  .........................................................................................................................
       Traceback (most recent call last):
         File "/Users/aphor/src/salt.git/tests/unit/transport/test_zeromq.py", line 335, in test_master_uri
           source_port=s_port) == 'tcp://[{0}]:{1};[{2}]:{3}'.format(s_ip, s_port, m_ip, m_port)
       AssertionError
   ......................................................................................................................................................................................
 ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
=========================================================================================================================================================================================
FAILED (total=7403, skipped=2100, passed=5298, failures=2, errors=3) 
================================================================================  Overall Tests Report  =================================================================================
```

isbm commented:
@aphor BTW, we are running all that with PyTest. :wink: JFYI.

aphor commented:
> @aphor BTW, we are running all that with PyTest. 😉 JFYI.

Trust, but verify @isbm ! :)

I think I will have some time this weekend to fix things up. I appreciate your attention!

rallytime commented:
@aphor Unfortunately this still isn't quite right. This is failing quite a few integration tests.

aphor commented:
IDK what is wrong with those tests. The failure logs say that kitchen didn't make a minion, and no targets matched :(

I'm going to fix stuff @isbm recommended, re-push, and then figure out what's wrong with the test kitchen.

cachedout commented:
@aphor I am re-running the test suite and pulling in the latest upstream branch changes.

aphor commented:
There's one unit test failure I found that I need to fix with salt.transport.zeromq._get_master_uri().

I tried running the daemon tests --docked, but there's something wrong with the test containers missing the salt test code, and I don't have time to figure out how to make that work so I'm going to just fix the master_uri bug and re-push.

aphor commented:
I know how to fix the tests now.

I need to handle an exception where the minion prep_ip_port() fails to parse the master as an ip_address (because it might be a hostname).

This is keeping the minion in any of the integration tests from running.

aphor commented:
This is encouraging: the tests are taking a while now, so it looks like the minion is no longer fail-fast.

I will rebase and re-push this PR if they go green, then remove the WIP.

aphor commented on salt/transport/zeromq.py:
```diff
@@ -69,33 +70,38 @@ def _get_master_uri(master_ip,
+                if master_ip_address.version == 4:
+                    ip_any = '0.0.0.0'
+                if master_ip_address.version == 6:
+                    ip_any = ip_bracket('::')
```
I understand what you're saying: I originally thought about that syntax also, but decided to do the more verbose form.

aphor commented:
The last round of testing passed for Ubuntu, but failed for Windows and CentOS on test infrastructure problems like kitchen failures and ssh tests failing which seem unrelated to this PR.

I think it's OK now.

cachedout commented:
This is looking a lot better. Once @isbm has a look I think we should be set to merge this.

aphor commented:
@isbm Bo, is there anything else I need to do to this PR to make you feel happy about it?

isbm commented on salt/minion.py:
```diff
@@ -243,27 +243,45 @@ def resolve_dns(opts, fallback=True):
+                    _port = _elements[1].split(':')[1]
+                except IndexError as exc:  # OK if there's no port specified, just pass
+                    log.debug("%s, unable to parse port number from opts['master'] IPv6 address: '%s'", exc, opts['master'])
+
```
Uhm... I am not sure I like it so much. Since now we have our own `ipaddress` that can address that, why not just enhance that and then reuse? Because this sort of parsing is a bit strange here.

aphor replied:
I agree, but it's not just an ip address. Documentation says ip:port or hostname:port combinations are supported. I agree it should be factored out, but I want to try factoring it out to a salt.utils.network.parse_host_port function.

Also, rather than writing a string parser, I'd like to try using regular expressions to separate the host and port elements.

aphor replied:
It needs to handle strings with hostname:port or ip:port combinations.

I factored it out of minion.py into salt.utils.network.parse_host_port and reworked the parsing to use Python regular expressions and ipaddress.

isbm commented on salt/minion.py:
```diff
@@ -243,27 +243,45 @@ def resolve_dns(opts, fallback=True):
+            try:
+                ret['master_port'] = int(_elements[1])
+            except IndexError as exc:  # OK if there's no port specified, just pass
+                log.debug("%s, unable to parse port number from opts['master'] IP address: '%s'", exc, opts['master'])
```
And this is another example why we should do it in `ipaddress` class, rather then here. We would not have this kind of duplication.

isbm commented on salt/transport/zeromq.py:
```diff
@@ -69,33 +70,42 @@ def _get_master_uri(master_ip,
+                if master_ip_address.version == 4:
+                    ip_any = '0.0.0.0'
+                if master_ip_address.version == 6:
+                    ip_any = ip_bracket('::')
```
Uhm, this looks a bit redundant. How about just:
```python
elif source_port and not source_ip:
    ip_any = '0.0.0.0' ipaddress.ip_address(master_ip).version == 4 else ip_bracket('::')
```

aphor replied:
OK. My inner Guido makes me believe the "bit redundant" is easier to read, but you win.

isbm reviewed (changes requested):
Sorry, I forgot all of that PR, so I re-reviewed it again. Here some suggestions...

rallytime commented:
Hi @aphor - There are some lint failures on this now. Can you fix those up?

https://jenkinsci.saltstack.com/job/pr-lint/job/PR-49755/30/warnings52Result/

aphor commented:
OMG hurry and merge it's all green in the Jenkins checks.... ;)
@isbm @rallytime

isbm commented on salt/utils/network.py:
```diff
@@ -1906,3 +1906,59 @@ def dns_check(addr, port, safe=False, ipv6=None):
     return resolved
+
+
+def parse_host_port(host_port):
```
Uhm... But do we really need all that? You also didn't wrote unit tests for it and yet you have a bug inside of the code. I don't really in favour of all this, to be honest. IPv6 ports are separated with `[` and `]`. So you don't need to maintain all these monstrous regular expressions, but just snip off the port away and let `ipaddress` does the rest. Something like this (untested, I wrote it directly here):

```python
from salt._compat import ipaddress

...

def get_host_port(hostport):
    host = port = None
    if not any([hostport.endswith(p) for p in ['.', ':']]):
        if ']:' in hostport:
            host, port = hostport[1:].split(']:')
        elif ':' in hostport and '.' in hostport:
            host, port = hostport.split(':')
        try:
            port = int(port)
        except ValueError, err:
            log.debug('Not a port: %s', err)
        try:
            host = ipaddress.ip_address(host)
        except ValueError, err:
            log.debug(err)

    return host, port
```
Look, ma! No regexes!

The rest `ipaddress` should just do. If it raises and exception "not an IPv6 or IPv4 address", then it is a host. Otherewise you will get an object instance where `.version` will tell you what it is.

isbm commented on salt/utils/network.py:
```diff
@@ -1906,3 +1906,59 @@ def dns_check(addr, port, safe=False, ipv6=None):
+        if _parser == 'ipv4':
+            s_to_a = lambda s: ipaddress.IPv4Address(s)  # pylint: disable=W0108
+        if _parser == 'ipv6':
+            s_to_a = lambda s: ipaddress.IPv4Address(s)  # pylint: disable=W0108
```
This is a bug.

isbm commented on salt/utils/network.py:
```diff
@@ -1906,3 +1906,59 @@ def dns_check(addr, port, safe=False, ipv6=None):
+    #   between the cases of host addressed as hostname, ipv4, or ipv6.
+    re_ipv4_port = re.compile('^(?P<host>\d{1,3}(?:\.\d{1,3}){3})(?:\:(?P<port>[0-9]{1,5}))?$')  # pylint: disable=W1401
+    re_ipv6_port = re.compile('^(?:\[(?P<host>(?:[0-9abcdef]{0,4}:)(?:(?:[0-9abcdef]{1,4})?:){1,6}[0-9abcdef]{0,4})\](?:\:(?P<port>[0-9]{1,5}))?|(?P=host))$')  # pylint: disable=W1401
+    re_hostname_port = re.compile('^(?P<host>(?:[a-zA-Z0-9]{1,62})(?:\.[a-zA-Z0-9]{1,62})*)(?:\:(?P<port>[0-9]{1,5}))?$')  # pylint: disable=W1401
```
Regex strings written as `r''` in Python.

aphor replied:
#### A, this is Moot:
if you say regular expressions are undesirable, and this can't pass review without it, then why complain about string type of regular expression constant?

#### B, lint:
You are too slow to review PRs to be nitpicking stuff like this. I'm a volunteer and my employer is not paying me for my contributions to salt. Please decide if this is important enough to enforce consistently or not and implement it as a lint test (or at least submit a ticket so that it's in the backlog and someone else can do it).

isbm replied:
@aphor I pointed bugs in your code, pointed it is unnecessarily hard to maintain and mentioned it has no enough unit tests coverage. I also helped you with an example how to do it simpler. But I will not participate provocative discussions, sorry.

isbm commented:
Summary:

- I am not in favour of the regexes when they are not even needed. I am totally against regexes when they are superhuge but not even needed.
- Unit tests are vastly missing.

aphor commented:
OK. I'll get rid of the regular expressions, but I don't like ad-hoc string parsing code strewn everywhere, and in this case the host:port syntax is a convention thats ~30 years old at least, meaning it's likely to pop up in places besides config['master'], ergo a good reuse candidate so I decided to refactor that out.

For some reason I thought it was already covered, but maybe because I'm coding after a full day of work and parental duties.. mistakes happen, I appreciate the oversight, and I'll add the coverage.

aphor committed:
coverage for master host:port ipv4 and ipv6 config value support
fix test_parse_host_port bad values
lint

aphor committed:
coverage for master host:port ipv4 and ipv6 config value support
fix test_parse_host_port bad values
lint

aphor committed:
fix parse_host_port() bad exception on hostname only arg

aphor committed:
fix parse_host_port() parse error on hostname only arg

aphor committed:
fix parse_host_port() parse error on hostname only arg

aphor committed:
fix parse_host_port() is_ip call

aphor committed:
avoid error on (redundant) is_ip check

aphor committed:
avoid error on (redundant) is_ip check

aphor committed:
pylint E741 triggers error in salt-pylint E8741 (unhandled)

s0undt3ch committed:
These should not be considered destructive tests

(cherry picked from commit 479439ab99a52e4bf0d8e56d6893a74baa57d5c3)

s0undt3ch committed:
Make sure blackout tests clean up after themselves. Properly.

(cherry picked from commit 9ab0e98e2c8f7f7cd15f06afcb2dedc940b4df26)

aphor committed:
debug ipaddress.ip_address TypeError

aphor committed:
debug ipaddress.ip_address TypeError

aphor committed:
fix indentation doh

aphor committed:
fix horrible mistakes

aphor committed:
lint pep8

aphor committed:
fix test setup

aphor committed:
fixup rebase merge goof

aphor committed:
test assertion correction

aphor committed:
fix good_host_ports iteration

aphor committed:
don't test invalid combination

aphor committed:
lint pep8 whitespace

aphor committed:
clear up lint, disambiguation

aphor committed:
typo in ipaddress.IPv6Address()

aphor committed:
checking for truthiness better than len

aphor committed:
remove bad extra test assertion inside exception handler

aphor committed:
don't try to handle/log test exception

aphor committed:
minimize diff

aphor committed:
parsing logic error

aphor committed:
wordsmithing

aphor committed:
expose error in tests

Error Message
object of type 'NoneType' has no len()
Stacktrace
Traceback (most recent call last):
  File "/tmp/kitchen/testing/tests/unit/utils/test_network.py", line 225, in test_parse_host_port
    host, port = network.parse_host_port(host_port)
  File "/tmp/kitchen/testing/salt/utils/network.py", line 1957, in parse_host_port
    raise _e_
TypeError: object of type 'NoneType' has no len()

aphor committed:
try to debug error in tests (revert me)

Error Message
object of type 'NoneType' has no len()
Stacktrace
Traceback (most recent call last):
  File "/tmp/kitchen/testing/tests/unit/utils/test_network.py", line 225, in test_parse_host_port
    host, port = network.parse_host_port(host_port)
  File "/tmp/kitchen/testing/salt/utils/network.py", line 1957, in parse_host_port
    raise _e_
TypeError: object of type 'NoneType' has no len()

aphor committed:
avoid TypeError by not constructing an ip_address from an ip_address

aphor committed:
ip_address does not need str() and parse_host_port handles ipv6 or ipv4

note:
@isbm requested ip_address objects stay ip_address. They implement __str__
so they should work without explicit str() conversion if necessary.

aphor committed:
Need more detail on test failure

aphor commented:
#36469 <-- this is popping up in unit test failures
`tests.unit.transport.test_zeromq.test_issue_36469_tcp`

I suspect, but don't know, that the Ubuntu 16.04 VM image is trying to connect to master via localhost, and is now resolving to '::1' IPv6 address and not '127.0.0.1' which may be an IPv6 problem in the OS image.

aphor commented:
> @aphor great! Except please do not in a future mention GitHub usernames in the commit messages. Because now every time this commit will be referenced/pushed/touched/transferred/etc, I will get another email. 😆

Sorry about that! I will amend and force push to get rid of that.

aphor commented:
@isbm I'd be grateful if you could review these changes.

I'm not sure why the py2-windows-2016 job is having problems. It doesn't look related to this change set.

aphor committed:
Don't ip_bracket addresses returned by check_dns.

This ends up in opts['master_ip'] where it can cause problems. I found
this while testng a git+pip based minion installation on Ubuntu Bionic.
I'm not sure why it doesn't affect other targeted OS versions in the
test framework. On Bionic, if 'salt' resolves to an IPv6 address,
zmq fails to connect to the master, and complains about the bracketed
address.

aphor committed:
syntax error bad parentheses

aphor committed:
socket.connect needs a tuple
TypeError
https://docs.python.org/2/library/socket.html
RTFD "Note This method has historically accepted a pair of parameters
for AF_INET addresses instead of only a tuple. This was never
intentional and is no longer available in Python 2.0 and later.
"

aphor committed:
Merge branch '2018.3' into issues/49754

aphor committed:
backport test improvements from develop
Bogus random test failures
forklifted from develop at 414bfe61a4fbe6f84be32242924f23c7126c24d6

aphor committed:
Merge branch 'issues/49754' of github.com:aphor/salt into issues/49754

aphor committed:
Revert "backport test improvements from develop"

This reverts commit 0af7653dfa3fc9b30601dec2c8ac2ce3cba52e7f.

I guess that was way too optimistic.

aphor commented:
tests.integration.minion.test_blackout is failing off and on for Ubuntu and Windows tests

I tried naïvely bringing the latest changes to that file from develop, but there's too much baggage.

@gtmanfred, @s0undt3ch is there anything you would recommend to make these red lights green?

s0undt3ch commented:
We have fixes in develop for that test which cannot be backported into 2018.3, but perhaps the 2019.2 branch changes are small enough to get that test more stable...

aphor commented on salt/utils/network.py:
```diff
@@ -56,12 +56,12 @@
 
 def sanitize_host(host):
-    '''
+    """
```
I'll fix that. I downloaded an Atom Python Black style enforcer, and it's been great for stuff I'm writing from scratch, but I haven't figured out how to get it to turn off by project :(

aphor committed:
fix docstring quote style

aphor committed:
wordsmithing

confession: tests failed to complete on gitub notification auth failure,
so I want to trigger another test.

aphor committed:
Revert "wordsmithing"

This reverts commit a07f3566516c3adb296e33fc0d3e2c9f30ac3eaf.

It wasn't necessary, and I want to trigger another round of tests on a Friday night.

aphor commented:
@rallytime I think this is ready for review.
